packages/back-end/src/routers/saved-queries/saved-queries.controller.ts

```diff
@@ -188,3 +205,227 @@ export async function deleteSavedQuery(
     status: 200,
   });
 }
+
+export async function postGenerateSQL(
+  req: AuthRequest<{ input: string; datasourceId: string }>,
+  res: Response
+) {
+  const { input, datasourceId } = req.body;
+  const context = getContextFromReq(req);
+  const { aiEnabled } = getAISettingsForOrg(context);
+
+  if (!orgHasPremiumFeature(context.org, "ai-suggestions")) {
+    throw new Error(
+      "Your organization's plan does not support generating queries"
+    );
+  }
+  if (!aiEnabled) {
+    return res.status(404).json({
+      status: 404,
+      message: "AI configuration not set or enabled",
+    });
+  }
+  if (!req.organization) {
+    return res.status(404).json({
+      status: 404,
+      message: "Organization not found",
+    });
+  }
+  const datasource = await getDataSourceById(context, datasourceId);
+  if (!datasource) {
+    return res.status(404).json({
+      status: 404,
+      message: "Datasource not found",
+    });
+  }
+  const informationSchema = await getInformationSchemaByDatasourceId(
+    datasource.id,
+    context.org.id
+  );
+  if (!informationSchema) {
+    return res.status(404).json({
+      status: 404,
+      message: "No informationSchema found for the datasource",
+    });
+  }
+  const maxTables = 100;
```

Comment:
This could get really expensive.  Most data warehouses are large and we will often hit this 100 limit.  It's also not uncommon for tables to have hundreds of columns, which might lead to a ton of token usage and unfocused results.

I wonder if a better approach is to do this in 2 steps.
1. Give only a list of table names (not columns) and have the AI return the ones it thinks it will need to answer the question.
2. Do the full prompt, but only include the tables and full schemas from step 1

The nice thing is we should already have the list of table names for every data source already.  We then only need to fill out missing schemas for a few tables at a time instead of blocking on all 100.

Reply:
I would suggest either always doing the 2-step process or setting this limit much lower, like 10.